### ISSUE TITLE
Improve readability of test_vectors.csv

### ISSUE THREAD

**Issue by siv2r** [ZhePang/Python_Specification_for_Schnorr_Adaptor#9]
The CSV file's *test type* and *result* columns are ambiguous, especially when we test the `adapt` and `extract_secadaptor` APIs. For instance,

- What does `TRUE` or `FALSE` represent for test type *Adapting* & *Extract Adaptor*?
    - `adapt`, and `extract_secadapt` APIs don’t output boolean values
- The reader has to dig through the code to find their meaning
    - For Adapting, `sig64 == sig_actual`
        - Shouldn't we instead use `schnorr_verify` to verify the output of the `adapt` API?
    - For Extract Adaptor, `t == adaptor`

Side Note 1: All the BIP340 test vectors were verified with only the `schnorr_verify` function. Therefore, the result column was the output of this function and didn’t need a *test-type* column. In our case, we require three different verification functions:

- `presig_verify` → tests validity of `extract_adaptor` result
- `schnorr_verify` → tests validity of `adapt` result
- `secadaptor_verify` → tests validity of `extract_secadaptor` result

Side Note 2: The API return types are inconsistent. `extract_adaptor` API (return type `Optional[Point]`) can return [1] Point, [2] None, or [3] False. However, `extract_secadaptor` & `adapt` will never return False. This inconsistency affects the value of the *result* column in the CSV table.

**Comment by siv2r** [comment]
I can think of the following solutions:

We can rename the “*test type”* column to “*target api”*  which will contain the APIs that the test vectors are used for

- soln 1:
    - make `adapt` and `extract_secadaptor` return False instead of `raiseValueError`
    - keep the result column unchanged
    - Pro: easy to implement
    - Cons: Confusing API return values
- soln 2:
    - make `extract_adaptor` return `[isValid, Point]` type instead
        - or we can use `raiseValueError` instead of returning False
        - or leave the return type as it is
    - the values of the result column would represent output values (VALID/INVALID) of the APIs present in the “*target api”* column
        - maybe add a side note (in the CSV file) about three different types of verification to better explain VALID/INVALID
    - Pros: test vectors linked to APIs, the user doesn’t need to look at Python code
    - Cons:
- soln 3:
    - implement auxiliary APIs `presig_verify`, `schnorr_verify`, and `secadaptor_verify` in reference.py
        - These API names will be values of the “*target api”* column
        - Their output will be the values of the *“result”* column
    - Pros: Detailed
    - Cons:
        - user needs to read the auxiliary API code to understand the CSV file
        - CSV file contains auxiliary API name (in _"target api"_ column) rather than main APIs (presign, adapt, etc.)
- soln 4:
    - don't change anything. Just add note in the CSV file for clarity.
        
I am leaning toward soln 2.

@jonasnick @ZhePang  What do you guys suggest?

**Comment by jonasnick** [comment]
> Side Note 2: The API return types are inconsistent. 

Yeah, there are multiple places where types are wrong:

```
❯ mypy test-vector.py
reference.py:143: error: Incompatible types in assignment (expression has type "tuple[int, int]", variable has type "bytes")  [assignment]
reference.py:144: error: Argument 2 to "point_add" has incompatible type "bytes"; expected "tuple[int, int] | None"  [arg-type]
reference.py:145: error: Argument 1 to "has_even_y" has incompatible type "tuple[int, int] | None"; expected "tuple[int, int]"  [arg-type]
reference.py:148: error: Argument 1 to "bytes_from_point" has incompatible type "tuple[int, int] | None"; expected "tuple[int, int]"  [arg-type]
reference.py:149: error: Argument 1 to "parity_from_point" has incompatible type "tuple[int, int] | None"; expected "tuple[int, int]"  [arg-type]
reference.py:149: error: Argument 1 to "bytes_from_point" has incompatible type "tuple[int, int] | None"; expected "tuple[int, int]"  [arg-type]
reference.py:151: error: Argument 2 to "schnorr_pre_verify" has incompatible type "bytes"; expected "tuple[int, int]"  [arg-type]
reference.py:175: error: Incompatible return value type (got "bool", expected "tuple[int, int] | None")  [return-value]
reference.py:179: error: Incompatible return value type (got "bool", expected "tuple[int, int] | None")  [return-value]
reference.py:184: error: Incompatible return value type (got "bool", expected "tuple[int, int] | None")  [return-value]
reference.py:194: error: Incompatible return value type (got "bool", expected "tuple[int, int] | None")  [return-value]
reference.py:295: error: Incompatible types in assignment (expression has type "tuple[int, int]", variable has type "bytes")  [assignment]
reference.py:296: error: Argument 2 to "schnorr_pre_verify" has incompatible type "bytes"; expected "tuple[int, int]"  [arg-type]
reference.py:368: error: Argument 1 to "compress_point" has incompatible type "tuple[int, int] | None"; expected "tuple[int, int]"  [arg-type]
reference.py:403: error: Missing return statement  [return]
reference.py:413: error: Argument 1 to "compress_point" has incompatible type "tuple[int, int] | None"; expected "tuple[int, int]"  [arg-type]
reference.py:427: error: Argument 1 to "compress_point" has incompatible type "tuple[int, int] | None"; expected "tuple[int, int]"  [arg-type]
test-vector.py:154: error: Argument 1 to "compress_point" has incompatible type "tuple[int, int] | None"; expected "tuple[int, int]"  [arg-type]
Found 18 errors in 2 files (checked 1 source file)
```

**Comment by jonasnick** [comment]
I agree that `extract_t`, `adapt` and `extract_adaptor` should not raise Errors except for the length of the inputs.

Side note: We should rename the functions and variables to be more understandable, similar what we've done in the C implementation.

We could also have separate CSV files for testing different API functions with sensible columns for the specific API tested.

**Comment by siv2r** [comment]
> We could also have separate CSV files for testing different API functions with sensible columns for the specific API tested.

I like this idea. This would simplify things.

How does four CSV files (one for each API) sound?
<h3>Presign</h3>

seckey | pubkey | message | aux_rand | adaptor | pre-signature | result | comment
-- | -- | -- | -- | -- | -- | -- | --

The result column is AND of the following boolean values:
- _pre-signature == adaptor_presign(sk, msg, adaptor, aux_rand)_
- _adaptor_presig_verify(pre-signature, adaptor, msg, pubkey) == TRUE_


----
<h3>Extract Adaptor</h3>

pubkey | message | pre-signature | adaptor | result | comment
-- | -- | -- | -- | -- | --

The result column is: _adaptor == extract_adaptor(pre-signature, msg, pubkey)_

----
<h3>Adapt</h3>

pre-signature | secadaptor | bip340 signature | result | comment
-- | -- | -- | -- | --

The result column is: _schnorr_verify(adapt(pre-signature, secadaptor))_

---
<h3>Extract Secadaptor</h3>

pre-signature | bip340 signature | secadaptor | result | comment
-- | -- | -- | -- | --

The result column is: _secadaptor == extract_secadaptor(pre-signature, bip340 signature)_

**Comment by siv2r** [comment]
> How does four CSV files (one for each API) sound?

We can also combine them into two CSV files instead: [1] Presign & Extract Adaptor [2] Adapt & Extract Secadaptor, as their columns are similar.

**Comment by siv2r** [comment]
Did this issue close automatically after #8 merged? If so, we need to reopen it. 

Initially, I had planned to fix this issue in the #8 pull request, but I decided against it to keep the PR brief. I will create a separate pull request for this.

**Comment by ZhePang** [comment]
Yeah I think it automatically closed it. I'll reopen it.
